Correct list-directed output of character strings

mohitdhingra22 opened the pull request:
Fixes: #8476
 - Added a single leading blank at the beginning of list-directed output.
- Ensured no extra blanks are inserted between adjacent character arguments.
- Added integration tests to confirm correct behavior.

Test program 
```fortran
program listdirected
  print *, '"','?','"'
end program listdirected
```
Before : 
```bash
"    ?    "
```
After :
```bash
 "?"
 ```

Pranavchiku reviewed (approved):
This looks towards good direction, feel free to merge once exhaustive checks pass.

HarshitaKalani reviewed (commented):
Hey @mohitdhingra22 , thanks for the PR. Can you please confirm if the leading spaces are expected? Otherwise the PR looks good.

mohitdhingra22 commented on tests/reference/run-array14-c9f6829.stdout:
```diff
@@ -1,2 +1,2 @@
-5.00000000e+00    -1.00000000e+00    3.00000000e+00    4.00000000e+00    2.00000000e+00
-5.00000000e+00    -1.00000000e+00    3.00000000e+00
+ 5.00000000e+00 -1.00000000e+00 3.00000000e+00 4.00000000e+00 2.00000000e+00
+ 5.00000000e+00 -1.00000000e+00 3.00000000e+00
```
The leading space in list-directed output is standard Fortran behaviour. It provides separation before printing values, especially for character strings, improving readability and ensuring consistent formatting across outputs. This matches what major Fortran compilers `(like gfortran)`do.

mohitdhingra22 commented:
> This should not print a leading space without `--print-leading-space`, and print the leading space with `--print-leading-space`.
 I have addressed the issue by modifying the formatting function to optionally print a leading space only when the `--print-leading-space` flag is provided. The default behavior now suppresses the leading space.
I updated relevant code, added tests, and verified the output with and without the flag. All tests are passing on my side.
Please let me know if you have any further suggestions.

kmr-srbh commented on src/libasr/runtime/lfortran_intrinsics.c:
```diff
@@ -1501,39 +1501,64 @@ void strip_outer_parenthesis(const char* str, int len, char* output) {
-    const int default_spacing_len = 4;
-    const char* default_spacing = "    ";
+    const int default_spacing_len = 1;
+    const char* default_spacing = " ";
```
```suggestion
    const char default_spacing = ' ';
```

kmr-srbh replied:
Let's keep the original change then.

kmr-srbh commented on src/libasr/runtime/lfortran_intrinsics.c:
```diff
@@ -1501,39 +1501,64 @@ void strip_outer_parenthesis(const char* str, int len, char* output) {
+    const char* default_spacing = " ";
     ASSERT(default_spacing_len == strlen(default_spacing));
-    *result = realloc(*result, result_capacity + 1 /*Null Character*/ );
+
+    *result = realloc(*result, result_capacity + 1);
```
Please add back the comment. It helps understand why the `+1` is needed. Although, I am not very sure if we even need that now, given we moved to null-independent strings sometime ago.

kmr-srbh replied:
@assem2002 could you please help with this?

kmr-srbh replied:
> I don't believe we need it, yes. But I think it's causing no harm for now. So let's keep it that way, Removing that `+1` could lead to some minor bugs that's totally unrelated to this PR.

Yeah, that's right! :)

kmr-srbh commented on src/libasr/runtime/lfortran_intrinsics.c:
```diff
@@ -1501,39 +1501,64 @@ void strip_outer_parenthesis(const char* str, int len, char* output) {
+        (*result)[1] = '\0';
+        result_size = 1;
+    } else {
+        (*result)[0] = '\0';
```
`result` is passed by reference and this value is already set inside `_lcompilers_string_format_fortran()` before calling `default_formatting()`. Could you please revert this change?

kmr-srbh replied:
`result` is passed by reference and this value is already set inside `_lcompilers_string_format_fortran()` before calling `default_formatting()`. Could you please revert this change?

kmr-srbh reviewed (approved):
The changes look good! I left some comments.

kmr-srbh commented on src/libasr/runtime/lfortran_intrinsics.c:
```diff
@@ -1587,7 +1612,7 @@ LFORTRAN_API char* _lcompilers_string_format_fortran(const char* format, int64_t
 
     if(format == NULL){
-        default_formatting(&result, result_size, &s_info);
+        default_formatting(&result, result_size, &s_info, false);
```
Looking closely, I think the function argument `print_leading_space` and its related handling is not actually required, since the leading space is handled as a separate string with `--print-leading-space`. 

@mohitdhingra22 could you please use this diff to remove the function argument:

```diff
diff --git a/src/libasr/runtime/lfortran_intrinsics.c b/src/libasr/runtime/lfortran_intrinsics.c
index 3ffa7b331..d7ded801f 100644
--- a/src/libasr/runtime/lfortran_intrinsics.c
+++ b/src/libasr/runtime/lfortran_intrinsics.c
@@ -1501,7 +1501,7 @@ void strip_outer_parenthesis(const char* str, int len, char* output) {
     }
 }
 
-void default_formatting(char** result, int64_t *result_size_ptr, struct serialization_info* s_info, bool print_leading_space){
+void default_formatting(char** result, int64_t *result_size_ptr, struct serialization_info* s_info){
     int64_t result_capacity = 100;
     int64_t result_size = 0;
     const int default_spacing_len = 1;
@@ -1513,11 +1513,6 @@ void default_formatting(char** result, int64_t *result_size_ptr, struct serializ
         fprintf(stderr, "Memory allocation failed\n");
         exit(1);
     }
-    if (print_leading_space) {
-        (*result)[0] = ' ';
-        (*result)[1] = '\0';
-        result_size = 1;
-    }
     bool prev_was_char = false;
 
     while(move_to_next_element(s_info, false)){
@@ -1546,7 +1541,7 @@ void default_formatting(char** result, int64_t *result_size_ptr, struct serializ
                 exit(1);
             }
         }
-        if(result_size > (print_leading_space ? 1 : 0) && !(prev_was_char && current_is_char)){
+        if(result_size > 0 && !(prev_was_char && current_is_char)){
             strcpy((*result) + result_size, default_spacing);
             result_size += default_spacing_len;
         }
@@ -1608,7 +1603,7 @@ LFORTRAN_API char* _lcompilers_string_format_fortran(const char* format, int64_t
     {fprintf(stderr,"Internal Error : default formatting error\n");exit(1);}
 
     if(format == NULL){
-        default_formatting(&result, result_size, &s_info, false);
+        default_formatting(&result, result_size, &s_info);
         free_serialization_info(&s_info);
         return result;
     }

```
